src/modules/commandrunner/src/lib/CommandRunner.cpp

```diff
@@ -1,465 +1,693 @@
-    if (nullptr != context)
-    {
-        char* name = reinterpret_cast<char*>(context);
-        if (FileExists(name))
+                Command::Status commandStatus = GetReportedStatus();
+                Command::Status::Serialize(writer, commandStatus);
+                status = CopyJsonPayload(payload, payloadSizeBytes, buffer);
+            }
+            else
+            {
+                OsConfigLogError(CommandRunnerLog::Get(), "Invalid object name: %s", objectName);
+                status = EINVAL;
+            }
+        }
+        else
         {
-            remove(name);
-            result = 1;
+            OsConfigLogError(CommandRunnerLog::Get(), "Invalid component name: %s", componentName);
+            status = EINVAL;
         }
     }
-    return result;
+
+    return status;
 }
 
-int CommandRunner::Cancel(std::string commandId)
+unsigned int CommandRunner::GetMaxPayloadSizeBytes()
 {
-    int status = MMI_OK;
-    CommandStatus* commandStatus = GetCommandStatus(commandId);
-    if ((nullptr != commandStatus) && (CommandState::Canceled != commandStatus->commandState))
-    {
-        OsConfigLogInfo(CommandRunnerLog::Get(), "Canceling command with CommandId: %s", commandId.c_str());
+    return m_maxPayloadSizeBytes;
+}
 
-        UpdatePartialCommandStatus(commandId, ECANCELED, CommandState::Canceled);
+void CommandRunner::WaitForCommands()
+{
+    m_commandQueue.WaitUntilEmpty();
+}
 
-        std::string tmpFilePath = GetTmpFilePath(commandStatus->GetUniqueId());
-        if (!FileExists(tmpFilePath.c_str()))
+int CommandRunner::Run(const std::string id, std::string arguments, unsigned int timeout, bool singleLineTextResult)
+{
+    std::shared_ptr<Command> command = std::make_shared<Command>(id, arguments, timeout, singleLineTextResult);
+    return ScheduleCommand(command);
+}
+
+int CommandRunner::Reboot(const std::string id)
+{
+    std::shared_ptr<Command> command = std::make_shared<ShutdownCommand>(id, "shutdown -r now", 0, false);
+    return ScheduleCommand(command);
+}
+
+int CommandRunner::Shutdown(const std::string id)
+{
+    std::shared_ptr<Command> command = std::make_shared<ShutdownCommand>(id, "shutdown now", 0, false);
+    return ScheduleCommand(command);
+}
+
+int CommandRunner::Cancel(const std::string id)
+{
+    int status = 0;
+    std::lock_guard<std::mutex> lock(m_cacheMutex);
+
+    if ((m_commandMap.find(id) != m_commandMap.end()) && !m_commandMap[id].expired())
+    {
+        std::shared_ptr<Command> command = m_commandMap[id].lock();
+        if (nullptr != command)
+        {
+            OsConfigLogInfo(CommandRunnerLog::Get(), "Canceling command with command id: %s", id.c_str());
+            status = command->Cancel();
+        }
+        else
         {
-            // Create empty temporary file
-            std::ofstream output(tmpFilePath);
-            output.close();
+            OsConfigLogError(CommandRunnerLog::Get(), "Command with command id %s is expired", id.c_str());

```

Comment:
>Command with command id %s is expired [](http://example.com/codeflow?start=55&length=37)

"Command %s is expired, cannot be canceled"

By the way, what really means "expired"?

May want to use another term.